[RDY] Building with VS2017

b-r-o-c-k opened the pull request:
This PR makes it possible to build Neovim with MSVC and Visual Studio 2017. After making changes to the CMakeLists and fixing some errors caused by missing headers, I was able to get it to compile successfully. MSYS2 isn't required anymore however, it could be used to install gperf which is the only dependency not downloaded by CMake. There are still runtime errors to fix, but I'd some feedback on my progress. 

Todo:
- [x] ~~Update LuaRocks after this [wget issue](https://github.com/luarocks/luarocks/issues/713) is resolved~~ A patch was added for LuaRocks, to make it use the curl.exe bundled inside the wintools.zip
- [x] Fix runtime errors
- [x] Check for regressions with other platforms/compilers
- [x] Add Appveyor CI for MSVC?
- [x] Update build documentation

b-r-o-c-k committed:
build/msvc: Fix linking for luajit and luv

b-r-o-c-k committed:
build/msvc: Fix missing restrict keyword

MSVC has the __restrict keyword and a marco is defined for it in `win_defs.h`.

b-r-o-c-k committed:
build/msvc: Remove unnecessary copy command for msgpack DLL

b-r-o-c-k committed:
build/msvc: Fix standard IO file number definitions

With MSVC, STDOUT_FILENO and STDERR_FILENO are defined as function calls instead of constants, meaning they can't be assigned to enum values. The enum was only used in one file, so it has been removed. A definition for STDIN_FILENO has been added that is consistent with the other two definitions.

b-r-o-c-k committed:
build/msvc: Fix preprocessor parsing

Preprocessor directives on the first line of the file were not being parsed.

b-r-o-c-k committed:
build/msvc: Remove unnecessary header include

b-r-o-c-k committed:
build/msvc: Allow building Unibilium on Windows

b-r-o-c-k committed:
build/msvc: Add MSVC compiler flag when building LuaRocks

b-r-o-c-k committed:
build/msvc: Add mode_t typedef to win_defs.h

b-r-o-c-k committed:
build/msvc: Fix name of import library

b-r-o-c-k committed:
build/msvc: Move include into unix_defs.h

b-r-o-c-k committed:
build/msvc: Use patched libvterm for MSVC

jamessan commented on third-party/CMakeLists.txt:
```diff
@@ -103,8 +103,13 @@ set(LUA_SHA256 2640fc56a795f29d28ef15e13c34a47e223960b0240e8cb0a82d9b0738695333)
-set(UNIBILIUM_URL https://github.com/mauke/unibilium/archive/v2.0.0.tar.gz)
-set(UNIBILIUM_SHA256 78997d38d4c8177c60d3d0c1aa8c53fd0806eb21825b7b335b1768d7116bc1c1)
+if(WIN32)
+  set(UNIBILIUM_URL https://github.com/b-r-o-c-k/unibilium/archive/v2.0.0.tar.gz)
```
Can you please try to get this merged upstream before asking us to pull from somewhere else?  I would suspect the header change to be fairly uncontroversial.  Not sure about the cmake side of things.

jamessan commented on third-party/CMakeLists.txt:
```diff
@@ -119,8 +119,13 @@ set(LIBTERMKEY_URL http://www.leonerd.org.uk/code/libtermkey/libtermkey-0.20.tar
 endif()
 
+if(MSVC)
+set(LIBVTERM_URL https://github.com/b-r-o-c-k/libvterm/archive/eb386b1d82f7d07363c9133b7aa06902ccd555fe.tar.gz)
```
Please try to work with upstream on MSVC support.  I thought VS 2015 had C99 support, so why don't VLA work?

b-r-o-c-k replied:
Although they have added support for some C99 features, they still don't support VLAs. The C90 standard is the only one they aim to be fully compliant with. You can read their response to the VLA feature request [here](https://visualstudio.uservoice.com/forums/121579-visual-studio-ide/suggestions/11580261-support-c99-variable-length-arrays).

jamessan reviewed (commented):
Thanks for the PR! Improving our Windows support is much appreciated.

My only comments so far are to avoid forcing us to carry patches for dependencies if at all possible.  Please work with the respective upstreams to try and fix the issues there.

If we need to carry patches, please do it using the existing mechanisms provided by ExternalProject_Add instead of having us pull from your forks.

jamessan commented:
@janlazo would probably be a good person to help review this.

b-r-o-c-k committed:
build/msvc: Add CMakeLists.txt and header for Unibilium

A header was added for compatibility with MSVC and CMakeLists.txt was
added for building with CMake.

b-r-o-c-k committed:
build/msvc: Add patch for libvterm

The patch removes VLAs because MSVC does not support them.

b-r-o-c-k committed:
build/msvc: Add patch for LuaRocks

LuaRocks bundles an outdated wget.exe for downloading packages on Windows. It is too old to support GitHub's TLS, so this patch will replace it with curl.

b-r-o-c-k committed:
build/msvc: Make libtermkey depend on Unibilium

b-r-o-c-k committed:
Merge branch 'master' into msvc-compat

b-r-o-c-k commented:
I contacted the libvterm developers and submitted a PR to Unibilium with the necessary patches. In case they don't get merged quickly, I also added the patches to this project.

zhaozg commented:
@b-r-o-c-k It's better if  appveyor build script.

justinmk commented on src/nvim/os/os_defs.h:
```diff
@@ -14,13 +13,6 @@
-  OS_STDIN_FILENO = STDIN_FILENO,
-  OS_STDOUT_FILENO = STDOUT_FILENO,
-  OS_STDERR_FILENO = STDERR_FILENO,
-};
```
@ZyX-I any objection to removing this?

ZyX-I replied:
This is not used in unit tests currently, so no objections. If it was then `enum` would be required.

justinmk commented on src/nvim/os/unix_defs.h:
```diff
@@ -1,9 +1,10 @@
+// Windows doesn't have the following headers,
+// so we include them here to avoid numerous instances of `#ifdef WIN32'.
 #include <unistd.h>
+#include <sys/param.h>
```
The comment is weird, is it still valid? `unix_defs.h` is only included by `os_defs.h`, specifically when `WIN32` is _not_ defined.

justinmk commented:
@b-r-o-c-k This is extremely appreciated and nicely done. Other than my comments LGTM and would like to merge it.  As @zhaozg mentioned we would want to quickly followup with a AppVeyor MSVC build, but in the meantime this is worth including.

 The ruby-related hang on AppVeyor can be investigated later (could be due to a upsteram change in the ruby client).

b-r-o-c-k committed:
build/msvc: Add missing WIN32 macro

MSVC predefines `_WIN32`, but not `WIN32`. Also, some unnecessary includes have been removed.

b-r-o-c-k committed:
build/msvc: Workaround for compiler optimization bug

b-r-o-c-k committed:
build/msvc: Make shell-test work

MSVC doesn't have unistd.h or usleep() so it was replaced with the
Sleep() WinAPI function.

b-r-o-c-k committed:
build/msvc: Remove pkgconfig dependency on Windows

b-r-o-c-k committed:
build/msvc: Add Appveyor CI for MSVC

b-r-o-c-k committed:
build/msvc: Fix errors caused by compiler intrinsics

b-r-o-c-k committed:
build/msvc: Remove confusing comment

b-r-o-c-k commented:
I added 64-bit and 32-bit MSVC build configs for AppVeyor. There are about 30 tests that are failing, but I think most of them are caused by it not being built with libintl.

justinmk commented on ci/build.bat:
```diff
@@ -1,27 +1,33 @@
+  C:\msys64\usr\bin\bash -lc "pacman --verbose --noconfirm -Su" || goto :error
+  C:\msys64\usr\bin\bash -lc "pacman --verbose --noconfirm --needed -S mingw-w64-%ARCH%-cmake mingw-w64-%ARCH%-perl mingw-w64-%ARCH%-diffutils mingw-w64-%ARCH%-unibilium gperf" || goto :error
+) else if "%CONFIGURATION:~0,4%" == "MSVC" (
+  set CMAKE_GENERATOR_ARGS=/verbosity:detailed
```
`/verbosity:detailed` looks like a 10x increase in log size. If we aren't debugging the build, let's use `/verbosity:normal` (which is the default, but leave it here explicitly so it's easy to change when needed).

justinmk commented on ci/build.bat:
```diff
@@ -42,19 +48,19 @@ where.exe neovim-node-host.cmd || goto :error
 cd build
-cmake -G "MinGW Makefiles" -DCMAKE_BUILD_TYPE=RelWithDebInfo -DBUSTED_OUTPUT_TYPE=nvim %USE_GCOV% -DGPERF_PRG="C:\msys64\usr\bin\gperf.exe" .. || goto :error
-mingw32-make VERBOSE=1 || goto :error
+cmake -G %CMAKE_GENERATOR% -DCMAKE_BUILD_TYPE=RelWithDebInfo -DBUSTED_OUTPUT_TYPE=nvim %USE_GCOV% -DGPERF_PRG="C:\msys64\usr\bin\gperf.exe" .. || goto :error
```
Looks like `C:\msys64\usr\bin\gperf.exe` is present on the default AppVeyor image.

b-r-o-c-k replied:
Yes, it is on the AppVeyor image, but I was wondering if we could add it to the [wintools.zip](https://github.com/neovim/deps/blob/master/opt/win32tools.zip) so Windows devs wouldn't need to download it themselves. Then the only requirements for building Neovim on Windows would be Visual Studio 2017, CMake, and git.

justinmk replied:
wintools.zip is for end-users, but we could instead host gperf.exe in the [deps repo](https://github.com/neovim/deps/tree/master/opt) and automatically download it to `.deps` with CMake.

justinmk commented on appveyor.yml:
```diff
@@ -1,7 +1,10 @@
+image: Visual Studio 2017
 configuration:
+- MSVC_64
+- MSVC_32
```
should be added to `allow_failures` below, then we can address the test failures in another PR.

justinmk commented:
@equalsraf any comments?

b-r-o-c-k committed:
build/msvc: Change MSBuild verbosity to normal

b-r-o-c-k committed:
build/msvc: Allow MSVC AppVeyor builds to fail

justinmk commented on test/functional/fixtures/shell-test.c:
```diff
@@ -4,7 +4,12 @@
 #include <stdint.h>
+#ifdef WIN32
+#include <Windows.h>
+#define usleep(usecs) Sleep(usecs/1000)
```
I think this should only be done `#ifdef _MSC_VER` . Maybe that's why the terminal tests fail on the mingw builds.

b-r-o-c-k replied:
Yes, I agree that it should be `_MSC_VER`, but I don't that it's causing any problems because the MinGW builds should be able to use all the Windows headers and functions.

justinmk commented on third-party/cmake/BuildLibtermkey.cmake:
```diff
@@ -1,5 +1,3 @@
-find_package(PkgConfig REQUIRED)
-
```
In `third-party/CMakeLists.txt` we pull from a different libtermkey repo:

```
if(WIN32)
set(LIBTERMKEY_URL https://github.com/equalsraf/libtermkey/archive/tb-windows.zip)
set(LIBTERMKEY_SHA256 c81e33e38662b151a49847ff4feef4f8c4b2a66f3e159a28b575cbc9bcd8ffea)
else()
...
```

to get this patch https://github.com/equalsraf/libtermkey/commit/b9317d6473b7a598359ebdfae5fd228bc159d0de

We could instead include that patch using `PATCH_COMMAND` as you've done with the others.

justinmk commented:
The MSVC build seems to have only 4 failed tests (nice!).  The `python3_spec` failures are probably an accident of https://github.com/neovim/python-client/issues/306, not related to this PR.

b-r-o-c-k committed:
build/msvc: Add Git as a requirement for Windows patches

b-r-o-c-k committed:
build/msvc: Make shell-test fix only apply to MSCV

Shougo commented:
> The MSVC build seems to have only 4 failed tests (nice!). The python3_spec failures are probably an accident of neovim/python-client#306, not related to this PR.

The change is reverted in Windows now.  So I think no problem.

justinmk commented:
Only thing left in your todo list is "Update build documentation", what did you have in mind? What steps are needed for a new contributor to get going with MSVC ?

Shougo commented:
lint is failed.  But it seems non related to the changes.

```
rc/nvim/eval.c:5943:  Extra space before ( in function call  [whitespace/parens] [4]
src/nvim/eval.c:5944:  Extra space before ( in function call  [whitespace/parens] [4]
src/nvim/option.c:2444:  /*-style comment found, it should be replaced with //-style.  /*-style comments are only allowed inside macros.  Note that you should not use /*-style comments to document macros itself, use doxygen-style comments for this.  [readability/old_style_comment] [5]
src/nvim/option.c:3312:  Small and focused functions are preferred: () has 696 non-comment lines (error triggered by exceeding 500 lines).  [readability/fn_size] [1]
Total errors found: 4
```

justinmk commented on third-party/CMakeLists.txt:
```diff
@@ -50,6 +50,13 @@ endif()
 option(USE_EXISTING_SRC_DIR "Skip download of deps sources in case of existing source directory." OFF)
 
+if(WIN32)
+  find_package(Git)
```
@b-r-o-c-k should `find_program` be used instead? https://cmake.org/cmake/help/v3.0/command/find_program.html

b-r-o-c-k replied:
`find_package` will use [FindGit.cmake](https://github.com/Kitware/CMake/blob/v2.8.7/Modules/FindGit.cmake) which is presumably better because it can find Git installations with non-standard names and paths.

justinmk commented on ci/build.bat:
```diff
@@ -1,27 +1,33 @@
   set ARCH=i686
   set BITS=32
-) else (
+) else if "%CONFIGURATION%" == "MINGW_64" (
```
This should stay as `else`. We need `ARCH` and `BITS` to be set for `MINGW_64-gcov`.

justinmk commented on ci/build.bat:
```diff
@@ -1,27 +1,33 @@
   set BITS=64
-)
-if "%CONFIGURATION%" == "MINGW_64-gcov" (
+) else if "%CONFIGURATION%" == "MINGW_64-gcov" (
```
Just `if`, not `else if`.

justinmk commented on ci/build.bat:
```diff
@@ -42,19 +48,19 @@ where.exe neovim-node-host.cmd || goto :error
 cd .deps
-cmake -G "MinGW Makefiles" -DCMAKE_BUILD_TYPE=RelWithDebInfo ..\third-party\ || goto :error
-mingw32-make VERBOSE=1 || goto :error
+cmake -G %CMAKE_GENERATOR% -DCMAKE_BUILD_TYPE=RelWithDebInfo ..\third-party\ || goto :error
```
should `%CMAKE_GENERATOR%` be in quotes?

b-r-o-c-k replied:
Right, but the quotes aren't implicit. Quotes are not needed on that line because they are part of the value assigned here:
```batch
set CMAKE_GENERATOR="Visual Studio 15 2017"
```
Batch is a very strange scripting language and it has mostly been made obsolete by PowerShell.

janlazo replied:
`set "CMAKE_GENERATOR=Visual Studio 15 2017"`

justinmk commented:
MSVC builds are failing with errors like this (probably libintl-related as mentioned):

```
Vim(language):E319: The command is not available in this version: lang C
...
Vim(language):E319: The command is not available in this version: lang mess C
```

Also noticed `gzip: Xtestfile.gz: Device or resource busy`.

## todo / future reference

- [HN thread](https://news.ycombinator.com/item?id=16524313) with info on generating PDB from mingw-w64 clang-generated executables. 
- Also use `PATCH_COMMAND` [for libtermkey](https://github.com/neovim/neovim/pull/8084#discussion_r172423013) #6659
- revisit d7ee3233264626de93e0466e0abecbc9dc6bc6ba
- revert b3d0241416407450056a18b2c603bdd48b7ceb9f after https://github.com/luarocks/luarocks/issues/713

b-r-o-c-k committed:
build/msvc: Fix AppVeyor build script for 'MINGW_64-gcov' configuration

b-r-o-c-k commented:
### Build Instructions for Visual Studio 2017
1. Install all of the following:
	- Visual Studio 2017 with the **Desktop development with C++** workload
	- CMake
	- Python 2 (required for libuv)
	- gperf
1. Build the dependencies (this will work in Command Prompt or PowerShell):
	```batch
	mkdir .deps
	cd .deps
	cmake -G "Visual Studio 15 2017" -DCMAKE_BUILD_TYPE=RelWithDebInfo -DPYTHON_EXECUTABLE="C:\path\to\python.exe" ..\third-party\
	cmake --build .
	```
1. Open the neovim folder in Visual Studio. This can be done from the file explorer using the "Open in Visual Studio" right-click menu option or from the menu in Visual Studio by selecting `File → Open → Folder`.
1. In the Solution Explorer, right-click the CMakeSettings.txt and select "Change CMake Settings". A new file named CMakeSettings.json will be created. Add this line
`"variables": [ {"name": "GPERF_PRG", "value": "C:\\path\\to\\gperf.exe"} ],`
below this line
`"name": "x86-Release",`
(This step can be skipped if gperf is in the PATH)
1. Select "x86-Release" configuration from the project settings dropdown and wait for CMake configuration to complete.
1. Click `CMake → Build All`.
Note: It is also possible to build with the "x64-Release" configuration, as long as `cmake -G "Visual Studio 15 2017 Win64"` is used to build the dependencies. However, the Debug configurations will not work because certain dependencies need to be linked with release version of the C runtime.
--------------------
In the future, we could make this easier by:
- Having the third-party/CMakeLists.txt download gperf in the same way it downloads other Windows executables.
- Removing the requirement for Python by including a CMakeLists.txt for libuv, such as the one used by [vcpkg](https://github.com/Microsoft/vcpkg/blob/master/ports/libuv/CMakeLists.txt).
- Making step 2 happen automatically when the CMake config detects it is running within Visual Studio. I think this would also remove the requirement for CMake because Visual Studio bundles it's own version of it.

justinmk commented on ci/build.bat:
```diff
@@ -1,27 +1,34 @@
+  :: These are native MinGW builds, but they use the toolchain inside
+  :: MSYS2, this allows using all the dependencies and tools available
+  :: in MSYS2, but we cannot build inside the MSYS2 shell.
+  set CMAKE_GENERATOR="MinGW Makefiles"
```
should be:

    set "CMAKE_GENERATOR=MinGW Makefiles"

justinmk commented:
Let's continue further work in other PR(s), this one is too big. Thanks @b-r-o-c-k !

justinmk commented:
## todo / future reference

MSVC builds are failing with errors like this (probably libintl-related as mentioned):

```
Vim(language):E319: The command is not available in this version: lang C
...
Vim(language):E319: The command is not available in this version: lang mess C
```

- Also noticed `gzip: Xtestfile.gz: Device or resource busy`.
- [HN thread](https://news.ycombinator.com/item?id=16524313) with info on generating PDB from mingw-w64 clang-generated executables. 
- Also use `PATCH_COMMAND` [for libtermkey](https://github.com/neovim/neovim/pull/8084#discussion_r172423013) #6659
- revisit d7ee3233264626de93e0466e0abecbc9dc6bc6ba
- revert b3d0241416407450056a18b2c603bdd48b7ceb9f after https://github.com/luarocks/luarocks/issues/713

oni-link commented on src/nvim/option.c:
```diff
@@ -2436,6 +2436,11 @@ static bool valid_filetype(char_u *val)
 
+#ifdef _MSC_VER
+// MSVC optimizations are disabled for this function because it
+// incorrectly generates an empty string for SHM_ALL.
```
The optimizer probably uses an empty string because the old pointer became invalid.

Problem seems to be that the use of `SHM_ALL` creates a local array that is assigned to pointer `p` (line 3201), but after leaving the block scope, `p` is used to access the (destroyed) array (line 3213).
